src/Select.js

```diff
@@ -569,30 +569,60 @@ var Select = React.createClass({
 			}
 		}
 
-		this.props.asyncOptions(input, (err, data) => {
-			if (err) throw err;
-			if (this.props.cacheAsyncResults) {
-				this._optionsCache[input] = data;
-			}
-			if (thisRequestId !== this._currentRequestId) {
-				return;
-			}
-			var filteredOptions = this.filterOptions(data.options);
-			var newState = {
-				options: data.options,
-				filteredOptions: filteredOptions,
-				focusedOption: this._getNewFocusedOption(filteredOptions)
-			};
-			for (var key in state) {
-				if (state.hasOwnProperty(key)) {
-					newState[key] = state[key];
+		if (typeof this.props.asyncOptions() !== 'undefined' &&
+			typeof this.props.asyncOptions().then === 'function') {
+			this.props.asyncOptions(input).then((data) => {
```

Comment:
> You call this.props.asyncOptions(input) and assign the result to a variable.

Meh, of course :sleeping: 

> But to be clear, you need to also support the callback case here,

Yep, I am calling it in https://github.com/dmatteo/react-select/blob/asyncOptions/src/Select.js#L594-L596, but I'm not passing it in the promise.